CFD-DEM Pneumatic Conveying Example (#1178)
chaos-polymtl/lethe: doc/source/examples/unresolved-cfd-dem/dense-pneumatic-conveying/dense-pneumatic-conveying.rst

The geometry of the pipe and the particle properties are based on the work of Lavrinec *et al*. [#lavrinec2021]_

---------------------------
Running the Simulations
---------------------------
Launching the simulations is as simple as specifying the executable name and the parameter file. Assuming that the ``lethe-particles`` and ``lethe-fluid-particles`` executables are within your path, the simulations can be launched in parallel as follows:

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-particles loading_particles.prm

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-particles settling_particles.prm

The particle loading and settling simulation should look like this:

.. raw:: html

    <p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/4uM51PCypZc?si=Xrisa4h87QLjvTWO" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

> 

Author: Good idea

.. code-block:: text
  :class: copy-button

>   mpirun -np 8 lethe-fluid-particles pneumatic-conveying.prm

Author: I have it, just need to add them I agree

The pneumatic conveying simulation should look like this:

.. raw:: html

    <p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/ESfSrmmlzYE?si=1RTsvFzcwvyelGme" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

.. note::
   The pneumatic conveying simulation lasts 5 seconds in this example, but last 10 seconds in the video. You can change the end time in the parameter file.

> Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_.

Author: I feel is nice to have a preview of the example before reading the parameter files, This section is pretty big in this example. :)

> Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_.

Reviewer: It's true that it is nice to have a preview like that before all the parameter file explanation... We should give some thought on maybe reformatting how we do our examples in general..

> Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_.

Reviewer 2: I understand wanting to give a preview, but I don't like that it changes the order of the section. If someones wants to see the results they can go directly to the results section in the table of content.
Let's keep the order of the sections the same for uniformity. Otherwise eventually all examples will have different names. I feel like scrolling down for the results is pretty easy and intuitive.

> Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_.

Author: Oki dokie

     set number of particle types = 1
     subsection particle type 0
       set size distribution type            = uniform
       set diameter                          = 0.005
       set number of particles               = 32194
       set density particles                 = 890
       set young modulus particles           = 1e6
       set poisson ratio particles           = 0.33
       set restitution coefficient particles = 0.3
       set friction coefficient particles    = 0.3
       set rolling friction particles        = 0.2
     end
     set young modulus wall           = 1e6
     set poisson ratio wall           = 0.33
     set restitution coefficient wall = 0.3
     set friction coefficient wall    = 0.4
     set rolling friction wall        = 0.2
   end

---------------------------------
Pneumatic Conveying Simulation
---------------------------------

The CFD simulation is to be carried out using the slug in the previous step. We will discuss the different parameter file sections.
The mesh and the DEM boundary condition sections are identical to the ones in the DEM simulations and will not be shown again.

Lagrangian Physical Properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The physical properties of the particles are the same as in the DEM simulations, except for the Young's modulus that was increased.

Author: These are the parameter value from the paper (10⁷), but I use 10⁶ in DEM so I can have a larger time step by the Rayleigh critical time step. I don't need a huge precision for the DEM simulations

The simulation lasts 5 seconds and the CFD time step is 5e-4 seconds.

.. code-block:: text

    subsection simulation control
      set method               = bdf1
      set output name          = cfd_dem
      set output frequency     = 10
      set time end             = 5
      set time step            = 5e-4
      set output path          = ./output/
    end

Physical Properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The physical properties of air are the same as Lavrinec *et al*. [#lavrinec2021]_

.. code-block:: text

    subsection physical properties
      subsection fluid 0
        set kinematic viscosity = 1.5e-5
        set density             = 1.205
      end
    end

Boundary Conditions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The boundary condition at the wall of the pipe is a weak function where the Dirichlet condition is weakly imposed as a no-slip condition, since we specify zero velocity. The inlet and the outlet have periodic boundaries. `See here <../../../parameters/cfd/boundary_conditions_cfd.html>`_ for more information about those boundary conditions.

.. code-block:: text

    subsection boundary conditions
      set number = 2
      subsection bc 0

>         set id   = 1

Reviewer 3: Should it be 0?

Author: Yes, I tested a few things and I forgot to change it back, nice catch!

.. code-block:: text

   subsection non-linear solver
     subsection fluid dynamics
       set solver           = inexact_newton
       set matrix tolerance = 0.1
       set reuse matrix     = true
       set tolerance        = 1e-4
       set max iterations   = 10
       set verbosity        = quiet
     end
   end

Linear Solver
~~~~~~~~~~~~~

.. code-block:: text

   subsection linear solver
     subsection fluid dynamics
       set method                                = gmres
       set max iters                             = 200
       set relative residual                     = 1e-4
       set minimum residual                      = 1e-6
       set ilu preconditioner fill               = 0
       set ilu preconditioner absolute tolerance = 1e-8
       set ilu preconditioner relative tolerance = 1.00
       set verbosity                             = quiet
       set max krylov vectors                    = 200
     end
   end

--------
Results
--------

The results presented here are obtained from a custom post-processing code that is currently not provided with the example.

Mass Flow Rate and Velocities
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Here we show the average velocities for the fluid, the slug and the particles in slug. The beta force, the averaged solid mass flow rate and the slug length over time are also shown. The shaded area represents the transient state.

Author: It is very approximate in this case since the simulation doesn't last long, this is when velocities fluctuate around the same value. I could have taken a more conservative start time (3.0 s) I agree. I could add a comment about that.

> Here we show the average velocities for the fluid, the slug and the particles in slug. The beta force, the averaged solid mass flow rate and the slug length over time are also shown. The shaded area represents the transient state.

Author: It was a good question, I did change it, 4.0s is better even tho we only have 1 s of data to process. This is fine for the example :)

==================================
Dense Pneumatic Conveying
==================================

It is strongly recommended to visit `DEM parameters <../../../parameters/dem/dem.html>`_  and `CFD-DEM parameters <../../../parameters/unresolved-cfd-dem/unresolved-cfd-dem.html>`_ for more detailed information on the concepts and physical meaning of the parameters used in the DEM and CFD-DEM solvers.

> 

Author: You are right, I tried to remove some redundancy, but I agree this one should be also removed!

.. code-block:: text

   subsection mesh
     set type                                = cylinder
     set grid type                           = balanced
     set grid arguments                      = 45 : 0.042 : 0.5
     set initial refinement                  = 1
     set expand particle-wall contact search = true
   end

.. note::
    Note that, since the mesh is cylindrical, ``set expand particle-wall contact search = true``. Details on this in the `DEM mesh parameters guide <../../../parameters/dem/mesh.html>`_.

A cross-section of the resulting mesh is presented in the following figure.

.. figure:: images/mesh-particle.png
    :alt: Mesh cross-section.
    :align: center

    Cross-section of the mesh used in the pneumatic conveying simulation.

Lagrangian Physical Properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The lagrangian properties were based on the work of Lavrinec *et al*. [#lavrinec2021]_, except for the Young's modulus that was deliberately reduced to get a smaller Rayleigh critical time step.

Author: oups

The gravity is set in the x-direction to allow the packing of the particles from the right side of the pipe.
The number of particles in the simulation is 32194. When the example was setup, the number specified in the simulation was higher since the insertion is done with the `plane insertion method <../../../parameters/dem/insertion_info.html#plane>`_, which will insert particles up to when they reach the plane.

> In order to avoid confusion with the number of particles in the parameter file, we did give the real number of particles inserted after 30 seconds.

Reviewer 2: Nah it's ok, if you feel it helps people understand lets keep it.

.. code-block:: text

    subsection restart
      set checkpoint = true
      set frequency  = 30000
      set restart    = false
      set filename   = dem
    end

-------------------
Settling Particles
-------------------

In this section we show the difference of the parameter file ``settling-particles.prm`` needed to settle the particles with the same gravity direction as the pneumatic conveying simulation. Also, many sections related to the loading are not needed such as the the insertion info, the floating walls, and the solid objects.

Simulation Control
~~~~~~~~~~~~~~~~~~~~
Here we allow a 2.5 seconds for the settling of the particles. Since this simulation is a restart of the loading particle simulation, the end time is 32.5 seconds.

.. code-block:: text

    subsection simulation control
      set time step        = 5e-5
      set time end         = 32.5
      set log frequency    = 500
      set output frequency = 1200
      set output path      = ./output_dem/
    end

Restart
~~~~~~~~

> This simulation reads the restart, meaning this option is set to true. Also, the checkpointing is reduce to 0.5 seconds.

Reviewer 2: ```suggestion
This simulation needs to read the restart files, consequently this option is set to true. Furthermore, the checkpointing ``frequency`` is reduced to 10 000 (every 0.5 seconds).
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The main difference in this simulation is the gravity. It changes to the y-direction to be coherent with the next simulation using the CFD-DEM solver.

.. code-block:: text

   subsection lagrangian physical properties
     set g                        = 0, -9.81, 0
     set number of particle types = 1
     subsection particle type 0
       set size distribution type            = uniform
       set diameter                          = 0.005
       set number of particles               = 32194
       set density particles                 = 890
       set young modulus particles           = 1e6
       set poisson ratio particles           = 0.33
       set restitution coefficient particles = 0.3
       set friction coefficient particles    = 0.3
       set rolling friction particles        = 0.2
     end
     set young modulus wall           = 1e6
     set poisson ratio wall           = 0.33
     set restitution coefficient wall = 0.3
     set friction coefficient wall    = 0.4
     set rolling friction wall        = 0.2
   end

---------------------------------
Pneumatic Conveying Simulation
---------------------------------

> The CFD simulation is to be carried out using the slug in the previous step. We will discuss the different parameter file sections.

Reviewer 2: ```suggestion
The CFD simulation is to be carried out using the slug generated in the previous step. We will discuss the different sections of the parameter file used for the CFD-DEM simulation.
```

The mesh and the DEM boundary condition sections are identical to the ones in the DEM simulations and will not be shown again.

Lagrangian Physical Properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The physical properties of the particles are the same as in the DEM simulations, except for the Young's modulus that was increased.

Author: ```suggestion
The physical properties of the particles are the same as in the DEM simulations, except for the Young's modulus that was increased to use the same value as the article [#lavrinec2021]_.
```

.. code-block:: text

   subsection lagrangian physical properties
     set g                        = 0, -9.81, 0
     set number of particle types = 1
     subsection particle type 0
       set size distribution type            = uniform
       set diameter                          = 0.005
       set number of particles               = 32194
       set density particles                 = 890
       set young modulus particles           = 1e7
       set poisson ratio particles           = 0.33

>        set restitution coefficient particles = 0.3

Reviewer 2: Why is the restitution coefficient so low? isnt that a bit weird? Generally it's around 0.9

Author: This is from the paper 
![image](https://github.com/chaos-polymtl/lethe/assets/71884806/3637508e-6c30-490e-9a1a-5d316e17838e)

Author: I will look into that

Author: They don't say anything about the COR. The material is HDPE.
This paper https://www.sciencedirect.com/science/article/pii/S0009250913005022#t0010 tested polyethylene pellets (they don't say if this is HDPE or LLDPE...) and they get results as low a 0.38. But the particles are not as spherical as the Lavrinec paper say for their particles.
Here they have a COR of 0.5 https://4spepublications.onlinelibrary.wiley.com/doi/epdf/10.1002/pen.20248, many papers that use HDPE cite them.

In my opinion, 0.3 is low, but not as far from 0.5. Honestly, I don't know how they got the value and I do not know if they calibrated the other parameters with it. They do cite a conference abstract about it, but I don't find it anywhere in the interwebs.

Reviewer 2: If its the same value as the paper, then don't worry. Just use the same values that they did. For a flow like this it won't have a huge impact anyway :)

Author: I used this value for my simulations, so I really wanted to be sure this was fine :) 
And since I looked at it, I prefer to put some info here than just say "this is from the paper" haha

       set friction coefficient particles    = 0.3
       set rolling friction particles        = 0.2
     end
     set young modulus wall           = 1e7
     set poisson ratio wall           = 0.33

>      set restitution coefficient wall = 0.3

Reviewer 2: Why is the restitution coefficient so low? isnt that a bit weird? Generally it's around 0.9

Author: Same as above

.. code-block:: text

    subsection physical properties
      subsection fluid 0
        set kinematic viscosity = 1.5e-5
        set density             = 1.205
      end
    end

Boundary Conditions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The boundary condition at the wall of the pipe is a weak function where the Dirichlet condition is weakly imposed as a no-slip condition, since we specify zero velocity. The inlet and the outlet have periodic boundaries. `See here <../../../parameters/cfd/boundary_conditions_cfd.html>`_ for more information about those boundary conditions.

.. code-block:: text

    subsection boundary conditions
      set number = 2
      subsection bc 0
        set id   = 0
        set type = function weak
        set beta = 100
        subsection u
          set Function expression = 0
        end
        subsection v
          set Function expression = 0
        end
        subsection w
          set Function expression = 0
        end
      end
      subsection bc 1
        set id                 = 1
        set type               = periodic
        set periodic_id        = 2
        set periodic_direction = 0
      end
    end

Flow control
~~~~~~~~~~~~~~~~~~

> Since the simulation has periodic boundary conditions, a correction force is needed to drive the flow to compensate the pressure drop in the pipe. In other to achieve this, we use the `dynamic flow controller <../../../parameters/cfd/dynamic_flow_control.html>`_. Here, we also apply a proportional beta force on particles. The average velocity is set to 3 m/s, this correspond to the average over the entire domain considering the void fraction. The flow controller performs well for CFD simulation, but needs some tuning for CFD-DEM simulation. By default, the controller has a high stiffness and aims to correct the flow in the next time step. However, the carrying of particles by the flow leads to a response time that is not taken into account and results in a oscillation of the velocity of the flow. To avoid this, we use the beta threshold and the alpha relaxation parameter. Here, the beta value will not be updated if the new beta value is within the 5% of the previous value. Also, the correction to apply to the previous beta value is reduced by a factor of 0.25. This way, the velocity of the flow and the particles are more stable.

Reviewer 2: ```suggestion
Since the simulation has periodic boundary conditions, a correction force is needed to drive the flow to compensate the pressure drop in the pipe. In other to achieve this, we use the `dynamic flow controller <../../../parameters/cfd/dynamic_flow_control.html>`_. Here, we also apply a proportional force (..math: `\beta`) on the particles. The average velocity is set to 3 m/s, this correspond to the average over the entire domain considering the void fraction. The flow controller performs well for CFD simulation, but needs some tuning for CFD-DEM simulation. By default, the controller has a high stiffness and aims to correct the flow in the next time step. However, the transport of particles by the flow leads to a response time that is not taken into account by the density of the gas and results in a oscillation of the velocity of the flow. To avoid this, we use the beta threshold and the alpha relaxation parameter. Here, the beta value will not be updated if the new beta value is within the 5% of the previous value. Also, the correction to apply to the previous beta value is reduced by a factor of 0.25. This way, the velocity of the flow and the particles are more stable.
```

I would use the symbol "beta" everywhere instead of saying beta. It's not like a "beta force" it's just a force that we call beta.

.. code-block:: text

   subsection flow control
     set enable               = true
     set enable beta particle = true
     set average velocity     = 3
     set flow direction       = 0
     set beta threshold       = 0.05
     set alpha                = 0.25
     set verbosity            = verbose
   end

Void Fraction
~~~~~~~~~~~~~~~

We choose the `quadrature centred method (QCM) <../../../theory/multiphase/cfd_dem/unresolved_cfd-dem.html#the-quadrature-centered-method>`_  to calculate the void fraction. The ``l2 smoothing factor`` we choose is the square of twice the diameter of the particles.

.. code-block:: text

    subsection void fraction
      set mode                = qcm
      set read dem            = true
      set dem file name       = dem
      set l2 smoothing factor = 0.0001
    end

CFD-DEM
~~~~~~~~~~

> The chosen drag model is Di Felice, and we use the Saffman lift force, the buoyancy force, and the pressure force. The coupling frequency is set to 100, which means that the DEM time step is 5e-6 s, for a Rayleigh critical time step of about 3.5%. The grad-div stabilization is used with a length scale of 0.084, the diameter of the pipe.

Reviewer: ```suggestion
The chosen drag model is Di Felice, and we use the Saffman lift force, the buoyancy force, and the pressure force. The coupling frequency is set to 100, which means that the DEM time step is 5e-6 s, for a Rayleigh critical time step of about 3.5 %. The grad-div stabilization is used with a length scale of 0.084, the diameter of the pipe.
```

> The chosen drag model is Di Felice, and we use the Saffman lift force, the buoyancy force, and the pressure force. The coupling frequency is set to 100, which means that the DEM time step is 5e-6 s, for a Rayleigh critical time step of about 3.5 %. The grad-div stabilization is used with a length scale of 0.084, the diameter of the pipe.

Reviewer 2: ```suggestion
We use the Di Felice drag model, the Saffman lift force, the buoyancy force, and the pressure force. The coupling frequency is set to 100, which means that the DEM time step is 5e-6 s. The DEM time-step is 3.5% of the Rayleigh critical time step. The grad-div stabilization is used with a length scale of 0.084, the diameter of the pipe.
```

Author: Fun fact, but I looked what is the right way to write "time-step" since I also though it was with a hyphen and the right ways are timestep or time step

.. code-block:: text

   subsection non-linear solver
     subsection fluid dynamics
       set solver           = inexact_newton
       set matrix tolerance = 0.1
       set reuse matrix     = true
       set tolerance        = 1e-4
       set max iterations   = 10
       set verbosity        = quiet
     end
   end

Linear Solver
~~~~~~~~~~~~~

.. code-block:: text

   subsection linear solver
     subsection fluid dynamics
       set method                                = gmres
       set max iters                             = 200
       set relative residual                     = 1e-4
       set minimum residual                      = 1e-6
       set ilu preconditioner fill               = 0
       set ilu preconditioner absolute tolerance = 1e-8
       set ilu preconditioner relative tolerance = 1.00
       set verbosity                             = quiet
       set max krylov vectors                    = 200
     end
   end

--------
Results
--------

> The results presented here are obtained from a custom post-processing code that is currently not provided with the example.

Reviewer 2: Is there a reason why the raw python files are not just included in the example? Even if they are not fully clean or are slightly hardcoded, I would still include them so someone can re-run the things.

Author: I talk about it in the PR description. 
It is because I stopped using the current version Lethe PyVista a while ago since new changes made my code and my output incompatible with it. Also, I use custom classes for my plots and for the processing of logs and PyVista files that are not specifically related to this example. 
My codes need a clean up, I don't feel confortable to just dump them as is in an example honestly. 
I wanna do a real clean up of my functions and put them in Utilities, since I have nice stuff, but I need to make it compatible with current tools and test it.

Reviewer 2: I understand, but then in the meantime I would remove the graphs from the results section then. The idea of the examples is that we include the tools necessary for the example to be re-run. If most of the graphs in the results cannot be remade from the example itself, then I don't think we should include them right now.
We can do this in a multiple stage way. Merge the example without the graphs that cannot be remade, and then once the utilities are put into place and the scripts can be made available, we add the results and the scripts.

Author: I can do a quick clean up for a standalone before the merge then. For me the fun part were the results, I perfer not to remove them :)

Reviewer 2: I agree, the results analysis you add they add a lot to the example.

Mass Flow Rate and Velocities
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Here we show the average velocities for the fluid, the slug and the particles in slug. The beta force, the averaged solid mass flow rate and the slug length over time are also shown. The shaded area represents the transient state. The quasi-steady state is approximated when velocities fluctuate around the same values.

Reviewer 2: ```suggestion
Here we show the average velocities for the fluid, the slug and the particles in slug. The beta force, the averaged solid mass flow rate and the slug length over time are also shown. The shaded area represents the transient state. The quasi-steady state is approximated when velocities fluctuate around the same values.
```

not sure about the "beta" force, I would refer it to it for a volumetric force (\beta) etc.

     subsection fluid dynamics
       set solver           = inexact_newton
       set matrix tolerance = 0.1
       set reuse matrix     = true
       set tolerance        = 1e-4
       set max iterations   = 10
       set verbosity        = quiet
     end
   end

-----------------------
Running the Simulations
-----------------------

Launching the simulations is as simple as specifying the executable name and the parameter file. Assuming that the ``lethe-particles`` and ``lethe-fluid-particles`` executables are within your path, the simulations can be launched in parallel as follows:

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-particles loading-particles.prm

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-particles settling-particles.prm

.. note::
   Running the particle loading simulation using 8 cores takes approximately 30 minutes and the particle settling simulation takes approximately 1 minute.

Once the previous programs have finished running, you can finally launch the pneumatic conveying simulation and get the simulation log for post-processing with the following command:

.. code-block:: text
  :class: copy-button

>   mpirun -np 8 lethe-fluid-particles pneumatic-conveying.prm | tee pneumatic-log.out

Reviewer 2: awesome :).

.. raw:: html

    <p align="center"><iframe width="800" height="450" src="https://www.youtube.com/embed/4uM51PCypZc?si=Xrisa4h87QLjvTWO" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

The pneumatic conveying simulation should look like this:

.. raw:: html

    <p align="center"><iframe width="800" height="450" src="https://www.youtube.com/embed/ESfSrmmlzYE?si=1RTsvFzcwvyelGme" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

.. note::
   The pneumatic conveying simulation lasts 5 seconds in this example, but last 10 seconds in the video. You can change the end time in the parameter file.

Post-processing
~~~~~~~~~~~~~~~
The data is extracted with the Lethe PyVista tool and post-processed with custom functions in the files ``pyvista_utilities.py`` and ``log_utilities.py``.

> Extraction, post-processing and plotting are automated in the script ``pneumatic-conveying_post_processing.py``:

Author: ```suggestion
Extraction, post-processing and plotting are automated in the script ``pneumatic-conveying_post-processing.py``:
```

.. code-block:: text
  :class: copy-button

>   python3 pneumatic-conveying_post_processing.py

Author: ```suggestion
  python3 pneumatic-conveying_post-processing.py
```
